### ISSUE TITLE
GtkApplication Root

### ISSUE THREAD

**Issue by ccoupe** [shoes/shoes3#400]
I thought I had an issue for this but I guess not. 

We don't use GtkApplication - but if we did - we could have menus - File  for Shoes apps (like Shoes/osx has) and like OSX - Linux/Gtk could send new shoes launches to the single Shoes process as new windows. OSX does this so the code is there in Shoes - part of the app vs APP[] thing. It's a bigger change you might think.  For example when you double click a .pdf file - you open the pdf reade app. If you double click a different pdf - it opens in the same reader. Web browsers do it  for html files.

**Comment by ccoupe** [comment]
This would also be a good place to add the multiple monitor support - they are related in odd ways.

**Comment by ccoupe** [comment]
Branch gapp (gtk) can request menus at Shoes.app time. Numerous and serious issues to resolve but lets get the api in order. @BackOrder cares about these things and it's better to get it right before a lot of code gets written. I offer:

```ruby
Shoes.app menus: true do
  mb = app.menubar
  # add Help and Help->About to menubar
  helpmenu = menu "Help"
  helpmenu << menuitem "About" do
   alert "Version 12 of Menu Test App"
  end
  mb << helpmenu
  # replace default File -> Quit handler  with our block
  filemenu = mb[0]
  filemenu.each do |fmi|
    if fmi.name == 'Quit' 
      fmi = menuitem "Musical Quit" do
        alert "I can't quit her - \"Blood, Sweat and Tears\". Use decoration"
      end
    end
  end
end
```
Feedback requested.

**Comment by dredknight** [comment]
Can I help with something?

**Comment by ccoupe** [comment]
@dredknight  - of course you can help with feedback. Thought Exercise, no coding for you, yet. 

The Cobbler screen has a _'button' panel_ - sort of toolbar-ish.  What if there was a real menubar on top with "File", "Cobbler" and the Cobbler menu had menu entries for "Info", "Gems".... with key accelerators like `control-i` and `control-g` . Obviously it would save visual space in cobbler - it has too many buttons already  and we need some more  Those could be menus and menuitems. It would extend Shoes into a very normal GUI toolkit for those that want it. 

I have a concerns about what self is in the {block} attached to the menuitem dispatch.

**Comment by ccoupe** [comment]
Do not mention the possibility of real tool-bar strips with ittty-bity icons  I need to keep that option open in this re-design.

**Comment by dredknight** [comment]
You are correct that cobbler can be arranged a bit more neatly. A menu-ish type thing sounds great.

For example (all mentions refer to latest 3.3.6 r3173 on windows) "Install gempack" and "Jailbreak gems"  should be inside "Install gems". It is intuitive that all gem related operations have to be there.

A cobbler bug I noticed is that if you selectg"Clear Image cache" twice (one after another) you get a C++ error.
Multiple errors appear in console  when  "Profile" button is selected.
```ruby
no implicit conversion of nil into String
```

**Comment by ccoupe** [comment]
Here is a small script and the screen shot.
```ruby
Shoes.app menus: true, width: 300, height: 200 do  
  mb = menubar
  helpmenu = menu "Help"
  aboutitem =  menuitem "About" do
    puts "ABOUT!!"
  end
  helpmenu << aboutitem
  mb << helpmenu
end
```
![menu2](https://user-images.githubusercontent.com/222691/36291736-07686ad6-128a-11e8-81ea-5fc7fd6526f2.png)

There are many issues - why are there scrollbars ? A known bug and wide ranging.  Who built the File Menu - the script didn't?   That requires that you know about Shoes OSX and the mac's global menu. On osx Shoes  has a global menu with File and and a few entries under it, Load and Quit.  When the hoes app close via the window decorator, that global menu is still there and Load will start a new app. People have hooked up their text editor via applescript to use it. All menu changes use the global menu on osx.  It can't be removed or changed from Shoes OSX. Just can't.

So, Shoes menus on Gtk (Linux and Windows) creates a File menu, and File->Quit for you, sort of like
OSX.  There is **a lot** to do. Next to do is a minimal OSX cocoa implementation so that little script works there.  Then I'll know enough to fill out the menu api for Shoes.

**Comment by ccoupe** [comment]
Now that I've looked at the OSX code for startup and menus there's a lot I'd like to change.  There are global menu items that haven't worked tor 4 years or more. Somethings that never worked and plenty of good intentions.  Much pondering.

**Comment by ccoupe** [comment]
I have a rough write up of the [menu api](https://github.com/shoes/shoes3/wiki/Menus)  Future work on the branch gapp will have to wait until I get out of the hospital and feel like working on Shoes - late March if all goes well.

**Comment by ccoupe** [comment]
I modified Cobbler to use the new menu system. Awesome Mockup. I am beyond pleased.  A  
![menu-cobbler](https://user-images.githubusercontent.com/222691/39666525-4bc1a114-5062-11e8-8721-f2ab9c080b95.png)

I added non functional menu items for the Advanced packaging (.deb) for me. This where I can integrate the long neglected exe-shoes code. Yay. And the linux-shoes and osx-shoes variations as well.  Sharp eyes might notice a menu for Theme.

**Comment by dredknight** [comment]
Looks great!

> Sharp eyes might notice a menu for Theme.
Does this mean I can put a colour or two to spice it up?

**Comment by BackOrder** [comment]
This is awesome and much better than the previous layout.

**Comment by ccoupe** [comment]
> Does this mean I can put a colour or two to spice it up?

Theme support is the next  feature I want to add to 3.3.7 - issue #354, #401 and others. It's CSS - if you know gtk css you can modify the appearance of lots of things. It's a fair sized task so it won't happen soon. There are lots of gtk themes that can be downloaded and installed on a linux system. I'm hoping to reuse them for Shoes. If so, Cobbler needs to browse them and switch to one. The screen shot is just a mockup for that. Lot's to do first. 

> This is awesome and much better than the previous layout.

I've tweaked Cobbler since. Menus  even more awesome when you see them  in action as the default for any app/script . [Download the 3.3.7  beta](https://walkabout.mvmanila.com/public/shoes/).  To set the default, create a `shoes.yaml` in the shoes root (next to the cshoes.exe for windows in the Program Files x86 place) 
```yaml
App_Name: Foobar
Theme: Sierra-light
Use_Menus: true 
```
Only the last line matters for current purposes. The 3.3.7 download also includes the multiple monitor stuff.

**Comment by ccoupe** [comment]
Does anyone need a small task to do in Ruby?  Merge packaging in Linux/Bsd uses the fpm command with is a ruby gem with a bin command line. It would be nice if Shoes could into the gem instead of having to having the user manual execute a fpm command line.  There are gotcha's to just shelling out to run the fpm command. [See here](https://github.com/shoes/shoes3/blob/gapp/lib/package/merge-fpm.rb#L207)

**Comment by dredknight** [comment]
I can do it but I am not sure that I understand what needs to be done. So FPM is a gem we have to go without using? From the code you have provided which are the functions/lines that need to be done with other tools?

**Comment by ccoupe** [comment]
The block comment is the code I want to make work. Call the gem insternals without going through
a shell call. 

I'm not sure fpm can be installed on Windows. On linux it installes the gem and then it installs a command in bin/ so it's called like anyother command line tool. the script in bin/ then calls the gem bin/fpm.rb (somehow - missing clarity).  which conspire to parse the command line and run the code in the gem.  The starting script is
```sh
#!/usr/bin/env ruby_executable_hooks
#
# This file was generated by RubyGems.
#
# The application 'fpm' is installed as part of a gem, and
# this file is here to facilitate running it.
#

require 'rubygems'

version = ">= 0.a"

if ARGV.first
  str = ARGV.first
  str = str.dup.force_encoding("BINARY") if str.respond_to? :force_encoding
  if str =~ /\A_(.*)_\z/ and Gem::Version.correct?($1) then
    version = $1
    ARGV.shift
  end
end

load Gem.bin_path('fpm', 'fpm', version)
```
which is kind of what 'require' does.

**Comment by dredknight** [comment]
FYI I built a centos environment this Friday where I will be testing the code. I will let you know once it is ready.

**Comment by ccoupe** [comment]
Closing 3.3.7 issues.
